pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,399 @@
+	<-stopCh
+}
+
+func (c *MCPodRouteController) gatewayWorker() {
+	for c.processGatewayNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processGatewayNextWorkItem() bool {
+	key, quit := c.gwQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gwQueue.Done(key)
+
+	if k, ok := key.(string); !ok {
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(k); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway(key string) error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	localStopCh := make(chan struct{})
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if activeGW == nil || c.nodeConfig.Name != activeGW.Name {
+		if c.started {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(localStopCh)
+			c.started = false
+		}
+		if len(c.podInfos) > 0 {
+			return c.deletePodFlows()
+		}
+		return nil
+	}
+
+	if c.nodeConfig.Name == activeGW.Name {
+		if !c.started {
+			klog.InfoS("Starting Multi-cluster PodFlowController")
+			localStopCh = make(chan struct{})
+			setPodInformer(c)
+			go c.podInformer.Run(localStopCh)
+			if !cache.WaitForNamedCacheSync(podRouteControllerName, localStopCh, c.podInformer.HasSynced) {
+				close(localStopCh)
+				return errors.New("failed to sync Pod cahce")
+			}
+			for i := 0; i < defaultWorkerNum; i++ {
+				go wait.Until(c.podWorker, time.Second, localStopCh)
+			}
+			c.started = true
+			return nil
+		}
+		// Do nothing when the Pod flow controller is already started since
+		// Pod flow controller will be responsible to handle Pod events to install flows.
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) podWorker() {
+	for c.processPodNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processPodNextWorkItem() bool {
+	obj, quit := c.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.podQueue.Done(obj)
+
+	if k, ok := obj.(types.NamespacedName); !ok {
+		c.podQueue.Forget(obj)
+		klog.InfoS("Expected types.NamespacedName in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncPod(k); err == nil {
+		c.podQueue.Forget(k)
+	} else {
+		c.podQueue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncPod(key types.NamespacedName) error {
+	pod, err := c.podLister.Pods(key.Namespace).Get(key.Name)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// The Pod has been removed, clean it up.
+			ips, exists := c.getPodNodeIPs(key)
+			if exists {
+				if len(c.ipToPods[ips.podIP]) > 1 {
```

Comment:
Needs locking too. I guess you need to lock the whole block, as another worker may be checking/processing the same IP, but then there is no difference from a single worker.

I remember earlier, you build the caches in enqueuePod(), and enqueue the Pod IP. What is wrong with that approach? In workers, you can just check the latest state of the IP from caches, and decide what to do - delete or add flows for the IP.

Reply:
Yeah, after more thinking about this, I added a go-routine background worker to clean up stale Pod flows after the podWorkers are stopped, since stale flows won't impact the traffic event it stays for a while. The cleanup job will run for every 5 seconds if podWorkers is not started and cache is not empty.

Reply:
You are talking about a different problem. For the race condition between `syncGateway` and `syncPod`, you can acquire a read lock of the `podWorkersStartedMutex` in `syncPod`, then `podWorders` can still run in parallel.

I am talking about multiple podWorders can get/modify `c.ipToPods` entry for the same IP, and based on the current value to decide what to do, so you need to lock the whole procedure of "get, checking value, do something, modify". For example, you can have the following unexpected sequence:

```
1. worker1: len(c.ipToPods[ips.podIP]) > 1
2. worker2: len(c.ipToPods[ips.podIP]) > 1
3. worker1: c.deleteIPToPods(key, ips.podIP)
4. worker2: c.deleteIPToPods(key, ips.podIP)

No one calls: c.deleteSinglePodFlow(key, ips.podIP)
```

Could you think about my suggestion:

> I remember earlier, you build the caches in enqueuePod(), and enqueue the Pod IP. What is wrong with that approach? In workers, you can just check the latest state of the IP from caches, and decide what to do - delete or add flows for the IP.

Let us chat if needed. I want you to really understand the problem.

Reply:
Thanks, got your idea here, there is already a `ipToPodsMutex` to protect `ipToPods`. I feel maybe we can add a mutex for the following belock?
```
				if len(c.ipToPods[ips.podIP]) > 1 {
					// A new Pod maybe assigned with the same IP and corresponding flow has been installed,
					// then no flow deletion action is needed.
					c.deleteIPToPods(key, ips.podIP)
				} else {
					return c.deleteSinglePodFlow(key, ips.podIP)
				}
```
For the previous implementation with IP in the queue, I feel it will be hard to determine which one is the latest one if any failure happened and re-queue. But If we get Pod in the queue, we can always get Pod's latest state via podLister.

Reply:
If you lock the block then:
>  I guess you need to lock the whole block, as another worker may be checking/processing the same IP, but then there is no difference from a single worker.

I did not get what you mean: "it will be hard to determine which one is the latest one if any failure happened and re-queue". `ipToPods` keeps the latest state right?

Reply:
Refined this part.